src/ring_theory/principal_ideal_domain.lean

```diff
@@ -175,6 +175,26 @@ begin
     (is_noetherian_ring.exists_factors a h)
 end
 
+/-- If a `ring_hom` maps all units and all factors of an element `a` into a submonoid `s`, then it
+also maps `a` into that submonoid. -/
+lemma ring_hom_factors_submonoid {R S : Type*}
```

Comment:
I don't think we need a `ring_hom` here, because the lemma as stated should be equivalent to `factors a ≤ f⁻¹ s -> units R ≤ f⁻¹ s -> a ∈ f⁻¹ s` (this is probably not the correct notation) and then we can drop the `f⁻¹` everywhere.

I'm also not sure about this name, perhaps something like `mem_submonoid_of_factors_subset_of_units_subset`?

Reply:
I don't understand. You mean the lemma should be `factors a ≤ s -> units R ≤ s -> a ∈ s`? But then we have to merge `R` and `S` into one. Then I won't be able to use the lemma any more where I need it.

Here is where I need it: `R` is the ring of polynomials. `S` is the type of linear maps. `f` is an evaluation homomorphism into linear maps. I need a lemmas that tells me that if a polynomial `a` is mapped by `f` to an invertible linear map, then all its factors are also mapped to invertible linear maps.

Reply:
Let me see if I can implement it, that should be easier for both of us :)

Reply:
Here is what I meant:

```lean
lemma mem_submonoid_of_factors_subset_of_units_subset (s : submonoid R)
  {a : R} (ha : a ≠ 0) (hfac : ∀ b ∈ factors a, b ∈ s) (hunit : ∀ c : units R, (c : R) ∈ s) :
  a ∈ s :=
begin
  rcases ((factors_spec a ha).2).symm with ⟨c, hc⟩,
  rw [← hc],
  exact submonoid.mul_mem _ (submonoid.multiset_prod_mem _ _ hfac) (hunit _),
end

/-- If a `ring_hom` maps all units and all factors of an element `a` into a submonoid `s`, then it
also maps `a` into that submonoid. -/
lemma ring_hom_mem_submonoid_of_factors_subset_of_units_subset {R S : Type*}
  [integral_domain R] [is_principal_ideal_ring R] [semiring S]
  (f : R →+* S) (s : submonoid S) (a : R) (ha : a ≠ 0)
  (h : ∀ b ∈ factors a, f b ∈ s) (hf: ∀ c : units R, f c ∈ s) :
  f a ∈ s :=
mem_submonoid_of_factors_subset_of_units_subset (s.comap f.to_monoid_hom) ha h hf
```

Reply:
Ok, makes sense. Thank you!